### DISCUSSION TITLE
Health Check Endpoint Change

### DISCUSSION THREAD

**Discussion by maheshyaddanapudi** [Netflix/conductor#2193] [General]
Hi,

It is noticed that the Healthcheck endpoint has been changed from /api/health to /health from 2.X to 3.X.
This would have some impact on existing Conductor users as their healthchecks and Web UIs (if any) would be pointing to the old url and will need a change.

Is there any reason why this was changed ?
Can it be reverted to old url without impact to any functionality being implemented in 3.X ?

Thanks and Regards,
Mahesh Yaddanapudi.

**Comment by MikeTangoEcho** [comment]
Hi,
another question following HealthCheck implementation:
* Is there someone already working on `spring-boot-actuator` integration in the 3.X ?
* With HealthCheck Indicator for each persistance ?

Regards

**Comment by kishorebanala** [comment]
This is an oversight, and the current health check only returns `true`. We intend to provide this as part of `spring-boot-actuator`, and through `/api/health`, without breaking the compatibility. If one of you have enough bandwidth to get to this before us, please let us know and send us a PR. Thank you!

**Reply by MikeTangoEcho** [reply]
Hi,
with only the gradle dependency `org.springframework.boot:spring-boot-starter-actuator`.

https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#actuator.enabling

I can reproduce the old behavior (with redis+elastic healthcheck) with the following properties in `config.properties`
```properties
# Change base path from /actuator to /api for backward compat
management.endpoints.web.base-path=/api
# management.endpoint.health.cache.time-to-live=5s
# management.endpoint.health.show-details=always

# Enable Health Indicator
management.health.elasticsearch.enabled=true
## Redis doesn't work since it doesn't use redis client as intended in the built-in HealthIndicator
## https://github.com/spring-projects/spring-boot/blob/v2.5.0/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/redis/RedisHealthIndicator.java
## Need a custom one
management.health.redis.enabled=true
management.health.jedis.enabled=true

# Disable other HealthChecks that Spring enable by default
management.health.diskspace.enabled=false
management.health.ping.enabled=false

# Force Enable Kubernetes tailored endpoints
## Spring <= 2.3.1
management.health.probes.enabled=true
## Spring > 2.3.1
management.endpoint.health.probes.enabled=true
management.health.livenessState.enabled=true
management.health.readinessState.enabled=true
## Add elastic and redis healthchecks to groups for k8s endpoint
management.endpoint.health.group.liveness.include=livenessState,elasticsearchRest,redis,jedis
management.endpoint.health.group.readiness.include=readinessState,elasticsearchRest,redis,jedis
```

Add the following class `JedisHealthIndicator` in `redis-persistence` along with `org.springframework.boot:spring-boot-starter-actuator` dependency if you need health check to be done on Redis connection.

```java
package com.netflix.conductor.redis.actuate;

import com.netflix.conductor.redis.config.AnyRedisCondition;
import com.netflix.conductor.redis.jedis.JedisProxy;
import org.springframework.boot.actuate.health.AbstractHealthIndicator;
import org.springframework.boot.actuate.health.Health;
import org.springframework.context.annotation.Conditional;
import org.springframework.stereotype.Component;

@Component
@Conditional(AnyRedisCondition.class)
public class JedisHealthIndicator extends AbstractHealthIndicator {

    protected JedisProxy jedisProxy;

    public JedisHealthIndicator(JedisProxy jedisProxy) {
        this.jedisProxy = jedisProxy;
    }

    @Override
    protected void doHealthCheck(Health.Builder builder) {

        try {
            jedisProxy.get("");
            builder.up();
        } catch (Exception e) {
            builder.withException(e)
                .down();
        }

    }

}
```

POST /api/health
```json
{
  "status": "UP",
  "components": {
    "elasticsearchRest": {
      "status": "UP",
      "details": {
        "cluster_name": "docker-cluster",
        "status": "yellow",
        "timed_out": false,
        "number_of_nodes": 1,
        "number_of_data_nodes": 1,
        "active_primary_shards": 16,
        "active_shards": 16,
        "relocating_shards": 0,
        "initializing_shards": 0,
        "unassigned_shards": 15,
        "delayed_unassigned_shards": 0,
        "number_of_pending_tasks": 0,
        "number_of_in_flight_fetch": 0,
        "task_max_waiting_in_queue_millis": 0,
        "active_shards_percent_as_number": 51.61290322580645
      }
    },
    "jedis": {
      "status": "UP"
    },
    "livenessState": {
      "status": "UP"
    },
    "readinessState": {
      "status": "UP"
    }
  },
  "groups": [
    "liveness",
    "readiness"
  ]
}
``` 

@kishorebanala Is there any plan to change the way redis is used for `org.springframework.data.redis.connection` so built-in Spring HealthIndicator works ? If not, do you prefer a dedicated module for each HealthIndicator implementation or one per persistance related module ?

**Reply by apanicker-nflx** [reply]
@MikeTangoEcho Thanks for providing an example, this is the implementation that we are considering as well using the spring boot actuators.

> Is there any plan to change the way redis is used for org.springframework.data.redis.connection so built-in Spring HealthIndicator works.

Currently, we do not plan to do this. 

>do you prefer a dedicated module for each HealthIndicator implementation or one per persistance related module

Ideally, it would be preferable to have an indicator per persistence module that can be enabled/disabled as desired.
